pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+		for ; batchSize <= maxBatchSize; batchSize = lo.Clamp(batchSize*2, 1, maxBatchSize) {
+			// add a timeout on increasing the batch size.
+			if c.clk.Since(multiBatchStart) > multiTimeout {
+				ConsolidationTimeoutTotalCounter.Inc()
+				break
+			}
+			maxIndex := lo.Clamp(i+batchSize, 0, len(candidates))
+			nodesToConsolidate := lo.Slice(candidates, i, maxIndex)
+			results, err := SimulateScheduling(ctx, c.kubeClient, c.cluster, c.provisioner, c.recorder, nodesToConsolidate...)
+			if err != nil {
+				return nil, fmt.Errorf("failed to simulate scheduling, %w", err)
+			}
+			valid, err := c.analyzeResults(ctx, results, nodesToConsolidate)
+			if err != nil {
+				return nil, fmt.Errorf("analyzing scheduling simulation results, %w", err)
+			}
+			// If not valid, don't double the batch size, and don't continue to simulate for this starting index
+			if !valid {
+				break
+			}
+			// If the command was valid, set the lastValidMax here
+			lastValidMax = maxIndex
+			// If we've reached the end of the list, we don't want to do another iteration of this loop
+			if maxIndex == len(candidates) {
+				break
+			}
+		}
+		// If we had a valid command created as part of the previous scheduling simulations,
+		// break the for loop and return that subset of candidates
+		if lastValidMax != -1 {
+			return candidates[i:lastValidMax], nil
+		}
+	}
+	// If we reach here, this means we never found a consolidation decision, so we return no candidates.
+	return nil, nil
+}
+
+// Candidates returns a list of the stateNodes ordered by their disruption cost
+func (c *Controller) Candidates(ctx context.Context) []*state.StateNode {
+	candidates := c.cluster.Nodes().Disruptable(ctx, c.kubeClient, c.recorder)
+	candidatesToPods := candidates.ReschedulablePodsForStateNodes(ctx, c.kubeClient)
+	nodePools := c.cluster.NodePools()
+	sort.Slice(candidates, func(i, j int) bool {
+		iDisruptionCost := disruptionutils.DisruptionCost(ctx, candidatesToPods[candidates[i]]) *
+			disruptionutils.LifetimeRemaining(c.clk, nodePools[candidates[i].Labels()[v1beta1.NodePoolLabelKey]], candidates[i].Node)
+		jDisruptionCost := disruptionutils.DisruptionCost(ctx, candidatesToPods[candidates[j]]) *
+			disruptionutils.LifetimeRemaining(c.clk, nodePools[candidates[j].Labels()[v1beta1.NodePoolLabelKey]], candidates[j].Node)
+		return iDisruptionCost < jDisruptionCost
+	})
+	return candidates
+}
+
+// Analyze results will look at the consolidation command, and return true if it should be executed
+//
+//nolint:gocyclo
+func (c *Controller) analyzeResults(ctx context.Context, results scheduling.Results, candidates []*state.StateNode) (bool, error) {
+	// if not all of the pods were scheduled, don't continue
+	if !results.AllNonPendingPodsScheduled() {
+		if len(candidates) == 1 {
+			c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, results.NonPendingPodSchedulingErrors())...)
+		}
+		return false, nil
```

Comment:
If we return back errors instead of valid/invalid here, we could put the event recording outside this function and just event anytime we get an invalid result 👀